algorithms/src/snark/marlin/ahp/prover/round_functions/first.rs

```diff
@@ -238,12 +255,61 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
 
         PoolResult::MatrixPoly(poly_for_opening, poly_for_committing)
     }
+
+    fn calculate_f<'a>(
+        label: impl ToString,
+        z_a: Vec<F>,
+        z_b: Vec<F>,
+        s_l: Vec<F>,
+        state: &prover::State<'a, F, MM>,
+    ) -> PoolResult<'a, F> {
+        let constraint_domain = state.constraint_domain;
+        // let v_H = constraint_domain.vanishing_polynomial();
+        // let should_randomize = MM::ZK && will_be_evaluated;
+        let label = label.to_string();
+        // let poly_time = start_timer!(|| format!("Computing {label}"));
+
+        let evaluations = z_a
+            .iter()
+            .zip(z_b.iter())
+            .zip(s_l.iter())
+            .map(
+                |((a, b), s)| {
+                    if s.is_zero() { F::zero() } else { *a + state.zeta * *b + (state.zeta.square() * *a * *b) }
+                },
+            )
+            .collect::<Vec<F>>();
+
+        let evals = EvaluationsOnDomain::from_vec_and_domain(evaluations, constraint_domain);
+
+        let poly = evals.interpolate_with_pc_by_ref(state.ifft_precomputation());
+        PoolResult::Query(LabeledPolynomial::new(label, poly, None, None))
+    }
+
+    fn calculate_selector<'a>(
+        label: impl ToString,
+        selector_evals: Vec<F>,
+        state: &prover::State<'a, F, MM>,
+    ) -> PoolResult<'a, F> {
+        let constraint_domain = state.constraint_domain;
+        // let v_H = constraint_domain.vanishing_polynomial();
+        // let should_randomize = MM::ZK && will_be_evaluated;
+        let label = label.to_string();
+        // let poly_time = start_timer!(|| format!("Computing {label}"));
+
+        let evals = EvaluationsOnDomain::from_vec_and_domain(selector_evals, constraint_domain);
+
+        let poly = evals.interpolate_with_pc_by_ref(state.ifft_precomputation());
+        PoolResult::Selector(LabeledPolynomial::new(label, poly, None, None))
+    }
 }
 
 #[derive(Debug)]
 pub enum PoolResult<'a, F: PrimeField> {
     Witness(LabeledPolynomial<F>),
     MatrixPoly(LabeledPolynomial<F>, LabeledPolynomialWithBasis<'a, F>),
+    Query(LabeledPolynomial<F>),
+    Selector(LabeledPolynomial<F>),
```

Comment:
Once we move the selector polynomial computation to indexing, we can get rid of this variant.